controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -9,515 +9,561 @@
+    CompletableFuture<Void> createEpochNode(int epoch) {
+        Preconditions.checkArgument(epochTxnMap.size() <= 2);
+        activeEpoch.compareAndSet(epoch - 1, epoch);
+        synchronized (epochTxnLock) {
+            epochTxnMap.putIfAbsent(epoch, new HashSet<>());
+        }
         return CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public CompletableFuture<Void> scaleOldSegmentsSealed(List<Integer> sealedSegments, List<Integer> newSegments, long timestamp) {
-        Preconditions.checkNotNull(sealedSegments);
-        Preconditions.checkArgument(sealedSegments.size() > 0);
-
-        List<List<Integer>> predecessors = new ArrayList<>();
-        List<SimpleEntry<Double, Double>> keyRanges = newSegments.stream().map(x -> {
-            Segment segment = FutureHelpers.getAndHandleExceptions(getSegment(x), RuntimeException::new);
-            return new SimpleEntry<>(segment.getKeyStart(), segment.getKeyEnd());
-        }).collect(Collectors.toList());
-        for (int i = 0; i < keyRanges.size(); i++) {
-            predecessors.add(new ArrayList<>());
+    CompletableFuture<Void> deleteEpochNode(int epoch) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        synchronized (epochTxnLock) {
+            if (epochTxnMap.getOrDefault(epoch, Collections.emptySet()).isEmpty()) {
+                epochTxnMap.remove(epoch);
+                result.complete(null);
+            } else {
+                result.completeExceptionally(new DataExistsException(""));
+            }
         }
-
-        int start = segments.size();
-        // assign status, end times, and successors to sealed segments.
-        // assign predecessors to new segments
-        for (Integer sealed : sealedSegments) {
-            InMemorySegment segment = segments.get(sealed);
-            List<Integer> successors = new ArrayList<>();
-
-            for (int i = 0; i < keyRanges.size(); i++) {
-                if (segment.overlaps(keyRanges.get(i).getKey(), keyRanges.get(i).getValue())) {
-                    successors.add(newSegments.get(i));
-                    predecessors.get(i).add(sealed);
-                }
+        return result;
+    }
+
+    @Override
+    CompletableFuture<Integer> createNewTransaction(UUID txId, long timestamp, long leaseExpiryTime, long maxExecutionExpiryTime,
+                                                    long scaleGracePeriod) {
+        Preconditions.checkNotNull(txId);
+
+        final CompletableFuture<Integer> result = new CompletableFuture<>();
+        final Data<Integer> txnData = new Data<>(
+                new ActiveTxnRecord(timestamp, leaseExpiryTime, maxExecutionExpiryTime, scaleGracePeriod, TxnStatus.OPEN)
+                .toByteArray(), 0);
+        activeTxns.putIfAbsent(txId.toString(), txnData);
+        int epoch = activeEpoch.get();
```

Comment:
this is expected. the caller is designed to handle data not found exception if an epoch gets deleted. 
this is same behaviour as we would get with zk. 
Take the following scenario:
Thread 1: gets the epoch. then tries to create txn node under epoch (This is non atomic)
Thread 2: tries to delete the empty epoch node. 

If thread 2 succeeds after thread 1 gets epoch but before it has successfully created txn node, thread 1 will get data not found exception. 
the caller handles that issue and retries.